src/library/type_context.cpp

```diff
@@ -3994,6 +3994,37 @@ static bool depends_on_mvar(expr const & e, buffer<expr> const & mvars) {
     return false;
 }
 
+buffer<bool> type_context_old::class_out_param_deps(expr const & cls) {
+    buffer<bool> result;
+    buffer<expr> C_args;
+    expr const & C = get_app_args(cls, C_args);
+    if (!is_constant(C))
+        return result;
+    expr it = infer(C);
+    for (unsigned int i = 0; i < C_args.size(); i++) {
+        if (!is_pi(it))
+            return result; /* failed */
+        expr const & d = binding_domain(it);
+
+        bool is_out = false;
+        // Is this parameter itself an out_param?
+        if (is_class_out_param(d)) {
+            is_out = true;
+        } else {
+            // Are there any bound variables in this parameter corresponding to an out_param?
+            for (unsigned int j = 0; j < i; j++) {
+                if (result[j] && occurs(mk_var(i - j - 1), d)) {
```

Comment:
```suggestion
                if (result[j] && has_free_var(d, i - j - 1)) {
```
Will also work under binders.

For bonus points add a `has_free_vars(d)` condition on the whole else-branch.